site/src/hooks/useClipboard.ts

```diff
@@ -1,44 +1,178 @@
+    // nowadays, but there's a fallback if not
+    if (typeof window?.navigator?.clipboard?.readText === "function") {
+      return {
+        success: true,
+        value: await window.navigator.clipboard.readText(),
+        error: null,
+      };
+    }
+
+    const { isExecSupported, value } = simulateClipboard("read");
+    if (!isExecSupported) {
+      throw new Error(
+        "document.execCommand has been removed for the user's browser, but they do not have access to newer API",
+      );
+    }
+
+    return {
+      success: true,
+      value: value,
+      error: null,
+    };
+  } catch (err) {
+    // Only expected error not covered by function logic is the user not
+    // granting the webpage permission to access the clipboard
+    const flattenedError =
+      err instanceof Error
+        ? err
+        : new Error("Unknown error thrown while reading");
+
+    return {
+      success: false,
+      value: null,
+      error: flattenedError,
+    };
+  }
+}
+
+// Comments for this function mirror the ones for readFromClipboard
+async function writeToClipboard(textToCopy: string): Promise<Result<void>> {
+  if (!document.hasFocus()) {
+    return {
+      success: false,
+      error: new Error(
+        "Security error - clipboard read queued while tab was not active",
+      ),
+    };
+  }
+
+  try {
+    if (typeof window?.navigator?.clipboard?.writeText === "function") {
+      await window.navigator.clipboard.writeText(textToCopy);
+      return { success: true, error: null };
+    }
+
+    const { isExecSupported } = simulateClipboard("write");
+    if (!isExecSupported) {
+      throw new Error(
+        "document.execCommand has been removed for the user's browser, but they do not have access to newer API",
+      );
+    }
+
+    return { success: true, error: null };
+  } catch (err) {
+    const flattenedError =
+      err instanceof Error
+        ? err
+        : new Error("Unknown error thrown while reading");
+
+    return {
+      success: false,
+      error: flattenedError,
+    };
+  }
+}
+
+type SimulateClipboardResult = Readonly<{
+  isExecSupported: boolean;
+  value: string;
+}>;
+
+function simulateClipboard(
+  operation: "read" | "write",
+): SimulateClipboardResult {
+  // Absolutely cartoonish logic, but it's how you do things with the exec API
+  const previousFocusTarget = document.activeElement;
+  const dummyInput = document.createElement("input");
+  dummyInput.style.visibility = "hidden";
+  document.body.appendChild(dummyInput);
+  dummyInput.focus();
+
+  // Confusingly, you want to use the command opposite of what you actually want
+  // to do to interact with the execCommand method
+  const command = operation === "read" ? "paste" : "copy";
+  const isExecSupported = document.execCommand(command);
+  const value = dummyInput.value;
+  dummyInput.remove();
+
+  if (previousFocusTarget instanceof HTMLElement) {
+    previousFocusTarget.focus();
+  }
```

Comment:
Decided to go with a much more "industrial-grade" approach compared to the previous code, and basically every other example I saw online

The biggest issue before was that when the dummy input received focus, there was nothing to put the focus back on the original element